### PR TITLE
Uncompressed string improvements

### PR DESCRIPTION
**Pull request by samansmink**
Adds to improvements to the uncompressed strings:
- gzips overflow strings that are written to disk
- no longer store string lengths but instead calculate them based on the dict offsets

Gzipping the big strings does have some significant impact on performance, but I think it's worth the lower disk space. Here's some microbenchmarks:

benchmark | with gzip | uncompressed
 --  | -- | -- 
storing big strings at 3x compression ratio | 1.98s | 0.50s
reading big strings at 3x compression ratio | 2.36s | 1.31s 

Note that this only holds for duckdb with persistent storage, big strings in memory are still stored uncompressed.

For the removal of the string lengths from the storage, I benchmarked scan performance but did not find any significant changes. There is a decent storage size improvement for small strings, to demonstrate this, I stored 1 million characters total divided over different string sizes:

string length | this PR | current master
-- | -- | --
1 | 5GB | 6.6GB
2 | 3GB | 4GB
4 | 2GB | 2.5GB
8 | 1.5GB | 1.7GB
16 | 1.2GB | 1.3GB

for (very) small strings, we get a nice reduction in size, which applies to both in-memory and persistent size.

### follow-up to this pr
Combining the removal of string lengths from this PR with fsst + delta encoding of offsets + bitpacking should give us very nice string compression

### CODE REVIEW

**Comment by Mytherin** [comment]
Thanks for the PR! Looks excellent. Although will it will have to wait until the feature-freeze is over to be merged.

@hannes and me have been discussing including `zstd` in the core engine, which should be considerably faster than gzip. This would also allow us to use `zstd` for extensions, which is a double whammy since it also gets a higher compression ratio than gzip at its highest setting.

**Comment by samansmink** [comment]
@Mytherin yea that sounds sensible, the overhead here is quite significant

**Review by Mytherin — APPROVED** [review]
LGTM, can be merged after the tests pass. Moving towards ZSTD can be a separate PR after this one.

**Pull request merged**
